cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java

```diff
@@ -357,9 +357,8 @@ protected void shutDown() throws Exception {
       try {
         if (mapReduce instanceof ProgramLifecycle) {
           destroy(success, failureInfo);
-        } else {
-          onFinish(success);
         }
+        onFinish(success);
```

Comment:
If you always `beforeSubmit` or `onFinish`, this mean when user is converting his MapReduce class, he has to be careful, otherwise it might get init or destroy twice, leading to error. E.g. He cannot just modify his class to call `beforeSubmit` from the `initialize` method.